mirrord/config/derive/src/config.rs

```diff
@@ -0,0 +1,105 @@
+use proc_macro2::{Span, TokenStream};
+use proc_macro2_diagnostics::{Diagnostic, SpanDiagnosticExt};
+use quote::{quote, ToTokens};
+use syn::{
+    spanned::Spanned, Data, DataStruct, DeriveInput, Fields, FieldsNamed, Ident, Visibility,
+};
+
+mod field;
+mod flag;
+
+use crate::config::{
+    field::ConfigField,
+    flag::{ConfigFlags, ConfigFlagsType},
+};
+
+#[derive(Debug)]
+pub struct ConfigStruct {
+    vis: Visibility,
+    ident: Ident,
+    fields: Vec<ConfigField>,
+    source: Ident,
+    flags: ConfigFlags,
+}
+
+impl ConfigStruct {
+    pub fn new(input: DeriveInput) -> Result<Self, Diagnostic> {
+        let DeriveInput {
+            attrs,
+            data,
+            ident: source,
+            vis,
+            ..
+        } = input;
+
+        let flags = ConfigFlags::new(&attrs, ConfigFlagsType::Container)?;
+
+        let fields = match data {
+            Data::Struct(DataStruct { fields, .. }) => match fields {
+                Fields::Named(FieldsNamed { named, .. }) => named
+                    .into_iter()
+                    .map(ConfigField::try_from)
+                    .collect::<Result<_, _>>()?,
+                _ => return Err(fields.span().error("Unnamed Structs are not supported")),
+            },
+            _ => return Err(source.span().error("Enums and Unions are not supported")),
+        };
```

Comment:
If we only care about 1 variant, think we should probably use `if let else` instead of `match`?

```suggestion
        let fields = if let Data::Struct(DataStruct { fields: Fields::Named(FieldsNamed { named, .. }), .. }) = data {
            Ok(named
                .into_iter()
                .map(ConfigField::try_from)
                .collect::<Result<_, _>>()?)
        } else {
            Err(source.span().error("Enums, Unions, and Unnamed Structs are not supported"))
        }?;
```